Add CombinedCodeSearchTool for unified code search

pranath-reddy opened the pull request:
### Summary :memo:
This PR adds the implementation of the `CombinedCodeSearchTool`, which aggregates results from multiple code search tools and reranks them using a pretrained `CrossEncoder` model. The combined tool queries the local semantic code search, GitHub keyword-based search (via SearxNG), and the SDE API-based search, returning a unified and ranked list of relevant code repositories.

### Details

- Introduced `CombinedCodeSearchTool` under `akd.tools.code_search` to support unified querying across multiple search tools.
- Combines results from:
  - `LocalRepoCodeSearchTool` for vector similarity search over local repositories
  - `GitHubCodeSearchTool` for keyword-based GitHub repository search via SearxNG
  - `SDECodeSearchTool` for querying the NASA SDE code search API
- Aggregates outputs and reranks them using a pretrained `CrossEncoder` model (e.g., `cross-encoder/ms-marco-MiniLM-L6-v2`)
- All results are returned in the standard `SearchResultItem` format with additional metadata:
  - `extra["tool"]` stores the source tool name
  - `extra["score"]` stores the reranker score assigned by the CrossEncoder
  
### Usage

```python
from akd.tools.code_search import (
    CodeSearchToolInputSchema,
    CombinedCodeSearchTool,
    CombinedCodeSearchToolConfig,
    SDECodeSearchTool,
    GitHubCodeSearchTool    
)

# Configure the tool
search_cfg = CombinedCodeSearchToolConfig()

# Initialize the code search tool
search_tool = CombinedCodeSearchTool(config=search_cfg, tools=[SDECodeSearchTool(), GitHubCodeSearchTool()])

# Run the search
result = await search_tool._arun(
    CodeSearchToolInputSchema(
        queries=["landslide nepal"],
        max_results=10
    )
)
```

### Bugfixes :bug: 
- Fixed a bug in `_sort_results` where `result["extra"]` was incorrectly used to access the `extra` field of `SearchResultItem`. It is now correctly accessed as `result.extra`.

pranath-reddy committed:
Combined Code Search Tool

- Added a combined tool that uses MS Marco MiniLM cross encoder for reranking
- Fixed a minor bug in _sort_results

NISH1001 commented on akd/tools/code_search.py:
```diff
@@ -139,6 +140,99 @@ def __get_sort_key(result):
+            except Exception as e:
+                logger.error(f"Error running tool {tool.__class__.__name__}: {e}")
+
+        reranked = self._rerank_results(all_results, params.queries[0])[: params.top_k]
```
Is it `params.top_k` or config? I think we should try something like:

```python
top_k = params.top_k or self.config.top_k
```

?

pranath-reddy replied:
it's `params.top_k`, we don't have a top_k field in the config schema

NISH1001 replied:
I mean max_results. THere's a computed_field as well.

pranath-reddy replied:
yes, `params.top_k` reflects `params.max_results` right?

NISH1001 replied:
Oh ya

NISH1001 commented on akd/tools/code_search.py:
```diff
@@ -139,6 +140,99 @@ def __get_sort_key(result):
+            SDECodeSearchTool(debug=debug),
+        ]
+        self.reranker_model = CrossEncoder(config.reranker_model_name)
+        super().__init__(config, debug)
```
Better and non-side effect will be following order

```python
...
super().__init(...)
self.tools = ...
self.reranker_model = ...
```

Notice: We invoke base's constructor first. This way we don't have to to `config = config or self.config_schema()` because super() will automatically handle it.

pranath-reddy committed:
Update code_search.py

- updated constructor to invoke base constructor first to avoid explicitly setting the config

pranath-reddy committed:
Update Code Search Input Schema

- added `top_k` as a computed field directly to `CodeSearchToolInputSchema`

NISH1001 commented on akd/tools/code_search.py:
```diff
@@ -232,17 +224,6 @@ def _rerank_results(
-    @computed_field
-    def top_k(self) -> int:
-        """Returns the number of top results to return."""
-        return self.max_results
```
Does this mean, these are removed, right?
